Spfx setup part2 - WSL (#2031)
pnp/blog: content/post/spfx-setup-part2-wsl/index.md

# Taxonomies
categories: ["Community post"]
tags: ["SharePoint Framework (SPFx)", "WSL"]
type: "regular"
---

## Intro

When starting your journey with SharePoint Framework (SPFx) development, you always start with a basic setup of your development environment. In this post series, I would like to introduce you to different ways to prepare for SPFx development. Over the next few posts, I will try to briefly explain how to:

- [perform local setup using Node Version Manager (NVM) or Node Version Switcher (NVS)](https://pnp.github.io/blog/post/spfx-setup-part1-nvm-nvs/) 
- boost local setup using WSL (which is this post)
- use Docker for SPFx development
- code in the cloud using GitHub Codespaces

## What is WSL?

> What if I told you that you can run a Linux on Windows without the need to setup any virtual machine using Virtualbox or VMWare or separate booting or anything like that? That's right! Windows Subsystem for Linux (WSL) allows you to install and run Linux (such as Ubuntu, Kali, Debian, or others) and use Linux applications, utilities, and Bash command-line tools directly on Windows. Yep just like that!

Reviewer: ```suggestion
What if I told you that you can run Linux on Windows without the need to setup any virtual machine using Virtualbox or VMWare or separate booting or anything like that? That's right! Windows Subsystem for Linux (WSL) allows you to install and run a Linux Distribution (such as Ubuntu, Kali, Debian, or others) and use Linux applications, utilities, and Bash command-line tools directly on Windows. Yep just like that!
```

## Why use WSL for your SPFx workspace setup and development?

> This is a good question. Why would you use WSL for your SPFx workspace setup and development? Well, I may name a few reasons:

Reviewer: ```suggestion
But why would you use WSL for your SPFx workspace setup and development? Well, I may name a few reasons:
```

```sh
wsl --shutdown
```

Ok but now let's get back to setting up our Ubuntu environment for SPFx development. First, let's log in again as the user we created:

```sh
wsl --distribution Ubuntu --user adam
```

Now we of course need to install Node.js and you may do it directly or you may use NVM or NVS for that. I will of course use a Node Version Manager (in this case NVM) to install Node.js, and if you are curious why please read the [previous part of this series](https://pnp.github.io/blog/post/spfx-setup-part1-nvm-nvs/). To install NVM you may run:

```sh
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.40.1/install.sh | bash
```

After that, we may install Node.js using NVM. For the currently latest version of SPFx which is 1.20 we need Node v18 so let's run:

```sh
nvm install 18.20.3
```

After that gets installed let's switch to that version by running: 

```sh
nvm use 18.20.3
```

Now we need to install the global dependencies for SPFx development, meaning Yeoman, Gulp, and the SharePoint Framework Yeoman generator. To do that you may run:

```sh
npm install gulp-cli yo @microsoft/generator-sharepoint --global
```

> And that's it! You are ready to start your SPFx development using WSL2. Remember that you may export your WSL environment at any time and import it when needed using the commands I mentioned at the beginning of this post. That way you may back up your setup, and share it with others or simply start from scratch when needed.

Reviewer: I think I would move the example with the export and import commands to this place here. In the other place the section is about reasons why to use wsl, and inserting script examples there is kind of odd. Also you only introduce the wsl command in the next chapter, so it does not make sense there as well.

Author: Good call 🤙

![comparing gulp bundle](images/comparing-gulp-bundle.png)

Now the result is significantly better on WSL2 than on Windows. I would say it is incomparable as in the case of Ubuntu on WSL2 we are working in seconds and in the case of Windows we are working in minutes. When we would run the `gulp bundle` command again, the result would be similar on both setups.
Now let's run the `gulp package-solution` command on both setups and see how long it will take to package the solution.

![comparing gulp package](images/comparing-gulp-package.png)

In this case, the result is quite similar but still slightly better on WSL2 than on Windows.

## VS Code support

VS Code has some pretty nice integrations when it comes to WSL. 

![remote development](images/remote-explorer.png)

By installing the Remote Development extension pack you may easily browse your WSL distros, open a terminal, and even open a full VS Code window in WSL. This is very useful when you want to work on your project in WSL but still use the full power of VS Code.

![remote development view](images/remote-explorer-view.png)

> Also after you connect, VS Code has and remote indicator that is presenting you are connected to WSL which all brings a quite smooth experience.

Reviewer: ```suggestion
Also after you connect, VS Code has a remote indicator that is presenting you as being connected to WSL which all brings a quite smooth experience.
```

![remote development indicator](images/remote-explorer-indicator.png)

Of course, when you just use the `code .` command from the  terminal in WSL, VS Code will open the project in WSL as well.

## SPFx Toolkit for Visual Studio Code

> When you are working on SPFx projects in VS Code, you may use the [SharePoint Framework Toolkit extension](https://marketplace.visualstudio.com/items?itemName=m365pnp.viva-connections-toolkit). This extension will help you at every stage of SharePoint Framework development. Doesn't matter if you are just starting your journey with SPFx or you are an experienced developer, this extension will help you from learning to boosting code, managing your solutions, and even deploying them to SharePoint. And now it comes along with an AI assistant dedicated for SharePoint Framework development. 

Reviewer: ```suggestion
When you are working on SPFx projects in VS Code, you may use the [SharePoint Framework Toolkit extension](https://marketplace.visualstudio.com/items?itemName=m365pnp.viva-connections-toolkit). This extension will help you at every stage of SharePoint Framework development. Doesn't matter if you are just starting your journey with SPFx or you are an experienced developer, this extension will help you from learning to boosting code, managing your solutions, and even deploying them to SharePoint. And now it comes along with an AI assistant dedicated to SharePoint Framework development. 
```
